### ISSUE TITLE
[TwigComponent] Render component in controller

### ISSUE THREAD

**Issue by Hugo-Le-Goff** [symfony/ux#2446]
Hello !

I am actually trying to render component in my symfony Controller/FormType to make a HTML label with one of my component. I achieved it by using `$this->renderView()` but when it comes to rendering components I didn't designed like the UX Icon one, i cannot use this trick.

I was wondering if adding a functionnality like it already exists with `InteractsWithTwigComponents` trait and its function `$this->renderTwigComponent()` was a good Idea ?

Beacause when it comes to having complex form labels it pretty hard to render them. Maybe I'm doing it the wrong way ?

**Comment by Kocal** [comment]
Hi, 

> $this->renderView() but when it comes to rendering components I didn't designed like the UX Icon one, i cannot use this trick.

I'm not sure to understand that part, you mean the DX can be improved? Or the rendered HTML is visually different?

I believe you can do something like that:
```php
$this->renderView('path/to/icon_template.html.twig', [
    'name' => 'flowbite:user-solid',
    'attributes' => ['height' => 24, 'width' => 24],
]);
```

```twig
<twix:ux:icon :name="name" {{ ...attributes }} />
```

Or did I miss something?

**Comment by Hugo-Le-Goff** [comment]
I'm indeed talking about DX, the rendered html is the same.

I just find calling the twig component template is not quite as elegant as a specific function like `renderTwigComponent`.

Moreover, when it comes to use dependencies components, your solution to create a twig template for it works fine but is redundant. Duplicating a template just to render it elsewhere looks unnecessary.

I was thinking of something like a little service to make this possible, it would make the work easier. Or even a built-in function in the controller maybe ? I don't really what it's possible to do.

**Comment by Kocal** [comment]
What about using [`ComponentRendererInterface#createAndRender()`](https://github.com/symfony/ux/blob/2.x/src/TwigComponent/src/ComponentRendererInterface.php#L19) ?

**Comment by smnandre** [comment]
>  in my symfony Controller/FormType

Not sure to follow here. Could you explain a bit more your use case ? 

Regarding UX Icons, you can directly inject the IconRendererInterface in your Controller/FormType.

But a generic way to render Twig component from controllers feels a bit odd to me right now ... I would at least need some use cases to understand.

**Comment by Hugo-Le-Goff** [comment]
Here is my use case, I have nice pills (with sometime icon inside) for the "access" value ("Accès" in french here) and I have a filter and research form at the top of my page. I want to render those pills with eventually icon inside as labels of this form.

![image](https://github.com/user-attachments/assets/010f26d2-932c-48ee-80b2-06024c0bd3bd)

What I did first was rendering the icon directly inside the template and pass the icon name as a parameter. And use renderView in my controller.

```twig
<span {{ attributes.without('class')}} class="{{ alert.apply({color, size}, attributes.render('class'))|tailwind_merge }}">
    {% if iconName is not null %}
        <twig:ux:icon name="{{ iconName }}"
                      class="size-2.5 {{ icon.apply({color, size}, attributes.render('class'))|tailwind_merge }}" />
    {% endif %}
    {% block content %}{{ content|raw }}{% endblock %}
</span>
```

It works, but I find 2 things wrong with that :
- I have to add a `content` parameter to write in the content block instead of defining it directly
- I cannot render UX Icon directly in the content

My main problem was not being able to call Icon. But with renderIcon I can, so thank you for the solution.

```php
->add('accessTypes', EnumType::class, [
   ...
    'label_html' => true,
    'choice_label' => function ($choice, string $key, mixed $value) use ($iconRenderer): string {
        return match ($choice) {
            BankAccessTypeEnum::PUBLIC => $this->renderView('components/Pill.html.twig', [
                'content' => $iconRenderer->renderIcon('fa6-solid:lock-open', ['class' => 'size-2.5']).'Public',
                'color' => 'green',
                'size' => 'md',
                'attributes' => new ComponentAttributes([]),
            ]),
            BankAccessTypeEnum::PRIVATE => $this->renderView('components/Pill.html.twig', [
                'content' => $iconRenderer->renderIcon('fa6-solid:lock', ['class' => 'size-2.5']).'Privé',
                'color' => 'red',
                'size' => 'md',
                'attributes' => new ComponentAttributes([]),
            ]),
        };
    },
])
```

For my own project I think I will stay on the first solution, beacause its simpler for my pills. 

But if someone else would like to do components rendering in the controller, with components inside another, with content block and other blocks. Wouldn't it be more convenient to have a something called `TwigComponentRendererInterface` with [this exact same syntax](https://symfony.com/bundles/ux-twig-component/current/index.html#test-helpers). So I could do something like that :

_(I understand that rendering the UX Icon with the twig component is not the best use case and that it's possible with IconRenderer, but imagine any other component than icon in this example)_

```php
$this->renderTwigComponent(
  name: 'Pill', 
  data: [
    'color' => 'green',
    'size' => 'md',
  ],
  content: $this->renderTwigComponent(
        name: 'UX:Icon',
        data: [
          'name' => 'fa6-solid:lock-open'
          'class' => 'size-2.5',
        ],
  .'Public', 
  blocks: [
      // Render others block
  ],
)
```

It would be much nicer to use, it would prevent from adding parameters to write in the blocks. Thus rendering any twig component would be as simple as calling this hypothetical interface.

Rendering components from a controller is something that can happens sometimes, so why not making it simpler and fully usable (with the blocks) ?

_Also there is nothing in the TwigComponents nor the UX Icon docs about rendering in the controller, it would be nice to have  a part about it._

I hope my point is clear now, maybe you'll find this useless but I think it's an interesting idea to improve the DX.

**Edit :**
I didn't see the message about the [ComponentRendererInterface#createAndRender()](https://github.com/symfony/ux/blob/2.x/src/TwigComponent/src/ComponentRendererInterface.php#L19). Nice to use but again nothing about the blocks. I think improving this with a content and a blocks parameter and mentionning it in the doc would be perfect.

I tried using it, one limitation, which is the same as renderView, is that it does not use the default values defined in the Twig\Components\\***.php file.

**Comment by smnandre** [comment]
I don't think we are going to push for from-controller usage for now... but who knows 🤷 

In your scenario, i would personaly not render labels like this but in a specific form theme template... (and/or probaly some cache management). 

Also as i see it, there is no scenario where a controller should/would need to only render a tiny component (like your badge).. and so i don't see what to do here :| 

But again i'm eager to get uses cases for this :)

**Comment by smnandre** [comment]
>  I think improving this with a content and a blocks parameter and mentionning it in the doc would be perfect

Blocks are working via Twig, in Twig.. it would not work if you called from your controller.

**Comment by Hugo-Le-Goff** [comment]
I understand your point of view, it's a niche use case and it can be done in others ways.

I'm happy that my idea was listened, and thankfull for the advices. I think you can close the discussion.

**Comment by smnandre** [comment]
Thank you for your input, and if you have any other feature idea, suggestion, or want to contribute in here, feel free to do so :)
